pkg/controllers/nodeclaim/disruption/controller.go

```diff
@@ -108,6 +99,84 @@ func (c *Controller) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (
 	return result.Min(results...), errs
 }
 
+type NodeClaimController struct {
+	*Controller
+}
+
+func NewNodeClaimController(clk clock.Clock, kubeClient client.Client, cluster *state.Cluster, cloudProvider cloudprovider.CloudProvider) corecontroller.Controller {
+	return corecontroller.Typed[*v1beta1.NodeClaim](kubeClient, &NodeClaimController{
+		Controller: NewController(clk, kubeClient, cluster, cloudProvider),
+	})
+}
+
+func (c *NodeClaimController) Name() string {
+	return "nodeclaim.disruption"
+}
+
+func (c *NodeClaimController) Reconcile(ctx context.Context, nodeClaim *v1beta1.NodeClaim) (reconcile.Result, error) {
+	return c.Controller.Reconcile(ctx, nodeClaim)
+}
+
+func (c *NodeClaimController) Builder(ctx context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1beta1.NodeClaim{}, builder.WithPredicates(
+			predicate.Or(
+				predicate.GenerationChangedPredicate{},
+				predicate.Funcs{
+					UpdateFunc: func(e event.UpdateEvent) bool {
+						oldNodeClaim := e.ObjectOld.(*v1beta1.NodeClaim)
+						newNodeClaim := e.ObjectNew.(*v1beta1.NodeClaim)
+
+						// One of the status conditions that affects disruption has changed
+						// which means that we should re-consider this for disruption
+						for _, cond := range v1beta1.LivingConditions {
+							if !equality.Semantic.DeepEqual(
+								oldNodeClaim.StatusConditions().GetCondition(cond),
+								newNodeClaim.StatusConditions().GetCondition(cond),
+							) {
+								return true
+							}
+						}
+						return false
+					},
+				},
+			),
+		)).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10}).
+		Watches(
+			&source.Kind{Type: &v1beta1.NodePool{}},
+			nodeclaimutil.NodePoolEventHandler(ctx, c.kubeClient),
+		).
+		Watches(
+			&source.Kind{Type: &v1.Node{}},
+			nodeclaimutil.NodeEventHandler(ctx, c.kubeClient),
+		).
+		Watches(
+			&source.Kind{Type: &v1.Pod{}},
+			nodeclaimutil.PodEventHandler(ctx, c.kubeClient),
+		),
+	)
+}
+
+type MachineController struct {
```

Comment:
Optional, parameterize the controller, in a single constructor e.g.

`NewController(v1alpha5.Machine{}, ...)`

Generics can remove a lot of this code, I think.

Looking a bit deeper, maybe this is the right approach, since your EventHandlers would then need to be parameterized.